### PR TITLE
batches: Create a batch change from the search result page

### PR DESCRIPTION
**Pull request by BolajiOlajide**
Closes [#22352](https://github.com/sourcegraph/sourcegraph/issues/22352)

## Test plan

* Run a search query on the search page
* Ensure you have the flag `experimentalFeatures.batchChangesExecution` set to true
* The `Create Batch Change` should be available on the search page, once clicked it'll redirect you to create a batch change with the batch spec already pre-filled.
CleanShot 2022-06-03 at 17 38 00@2x
 

## App preview:

- [Web](https://sg-web-bo-implement-batch-change-from.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-tareuoxmiy.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/web/src/enterprise/batches/batch-spec/yaml-util.ts [hunk 1]
```diff
@@ -383,32 +383,41 @@ export function quoteYAMLString(value: string): string {
...
  * `librarySpec`.
  *
  * @param librarySpec the raw batch spec YAML example code from a library spec
- * @param name the name of the batch change to be inserted
+ * @param value the value of the field to be updated
+ * @param key the name of the field in the spec to be updated
+ * @param quotable indicates if the value can be quoted or not
  */
-export const insertNameIntoLibraryItem = (librarySpec: string, name: string): string => {
+export const insertFieldIntoLIbraryItem = (
```
Comment (limitedmage) [inline client/web/src/enterprise/batches/batch-spec/yaml-util.ts:395]:
```suggestion
export const insertFieldIntoLibraryItem = (
```

File: client/web/src/search/results/SearchResultsInfoBar.tsx [hunk 2]
```diff
@@ -210,6 +214,12 @@ export const SearchResultsInfoBar: React.FunctionComponent<
     const createActions = useMemo(
         () =>
             [
+                getBatchChangeCreateAction(
+                    props.query,
+                    props.query as string,
```
Comment (limitedmage) [inline client/web/src/search/results/SearchResultsInfoBar.tsx:219]: Why is query being passed twice here? The second parameter should be the pattern type, not the query again.

File: client/web/src/search/results/createActions.ts [hunk 3]
```diff
@@ -78,3 +79,24 @@ export function getCodeMonitoringCreateAction(
...
+    query: string | undefined,
+    patternType: string,
+    authenticatedUser: Pick<AuthenticatedUser, 'id'> | null,
+    isServerSideBatchChangeEnabled: boolean | undefined
+): CreateAction | null {
+    if (!isServerSideBatchChangeEnabled || !query || !authenticatedUser) {
+        return null
+    }
+    const searchParameters = new URLSearchParams(location.search)
+    searchParameters.set('trigger-query', `${query} patterntype:${patternType}`)
```
Comment (limitedmage) [inline client/web/src/search/results/createActions.ts:93]: Do we also need to check that `patternType` is valid/non-empty? Also, I don't see the pattern type at all in the yaml spec when creating the batch change from the button, is this expected? (disclaimer: I've never used batch changes).

CleanShot 2022-06-03 at 11 21 06

**Review by limitedmage — COMMENTED** [review]
Looks good from the search code side, with some comments/questions.

Another minor UI issue I found: The placeholder text for the batch change name is an invalid example as it has spaces. The placeholder should be a valid example.

File: client/web/src/search/results/SearchResultsInfoBar.tsx (hunk 2 above)
Comment (BolajiOlajide) [reply client/web/src/search/results/SearchResultsInfoBar.tsx:219]: Dang! This is a case of rebasing gone wrong haha.
I'll fix this now.

File: client/web/src/search/results/createActions.ts (hunk 3 above)
Comment (BolajiOlajide) [reply client/web/src/search/results/createActions.ts]: Yes, it's not needed. I forgot to pus the final update that took care of that.

Comment (limitedmage) [reply client/web/src/search/results/createActions.ts]: Wait, so pattern type is not needed for batch changes? That seems odd. I would assume that a search with regex vs literal would result in significantly different search results.

File: client/web/src/enterprise/batches/batch-spec/edit/library/LibraryPane.tsx [hunk 4]
```diff
@@ -94,13 +95,36 @@ export const LibraryPane: React.FunctionComponent<React.PropsWithChildren<Librar
...
+            return updatedCode
+        },
+        [searchQuery]
+    )
+
     const onConfirm = useCallback(() => {
         if (selectedItem && !('isReadOnly' in props && props.isReadOnly)) {
-            const codeWithName = insertNameIntoLibraryItem(selectedItem.code, name)
-            props.onReplaceItem(codeWithName)
+            const updatedCode = updateLibraryItemFields(selectedItem.code, name)
```
Comment (courier-new) [inline client/web/src/enterprise/batches/batch-spec/edit/library/LibraryPane.tsx:123]: This is cool, but I also don't know how necessary/useful it is? I'm not sure. The examples each have different search queries that are more specific to each given example and that might not work terribly well if we drop in a totally different random search query. 😅 I think we could create a better experience with this come the library MVP/build-a-batch-spec, but for now I might be tempted to leave this part alone and _not_ replace the repo query.

**Review by courier-new — COMMENTED** [review]
This is looking great! However, it strikes me that we're kinda doing the same thing two different ways now. 😅 For creating a batch change from a code insight, we create a function that renders a batch spec template for the specific insight type ([this](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/enterprise/batches/create/go-checker-templates.ts?L643%3A1=)), get the template renderer on the Create page based on the URL parameter ([this](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/enterprise/batches/create/useInsightTemplates.ts?L22%3A5-24%3A71=)), and then update the batch spec on the backend immediately after the `create` mutation before we even navigate to the Edit page ([this](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/enterprise/batches/create/ConfigurationForm.tsx?L112%3A13-124%3A15=)). For creating a batch change from search this way, we're substituting the repo query field in the batch spec client-side from the Edit page at the time that the monaco editor first renders the default hello world batch spec.

Is there a way we could consolidate these two so it's a bit more consistent how we pre-generate a batch spec given a different entry point? I suppose I'm slightly partial to the code insights way because it means your search query actually gets persisted to the backend; right now if I create a batch change from search, leave the page, and come back via the batch change list page, my search query is gone.

File: client/web/src/search/results/createActions.ts (hunk 3 above)
Comment (eseliger) [reply client/web/src/search/results/createActions.ts]: The pattern type _is_ needed, but it defaults to whatever search defaults to. When the user selects any other pattern type, it must be reflected here.

Comment (BolajiOlajide) [reply client/web/src/search/results/createActions.ts]: Got it. Thanks - I'll make the update.

File: client/web/src/enterprise/batches/batch-spec/edit/library/LibraryPane.tsx (hunk 4 above)
Comment (BolajiOlajide) [reply client/web/src/enterprise/batches/batch-spec/edit/library/LibraryPane.tsx:117]: Got it. 

My thinking to updating that part of the code is that users might want to try a specific non-`Hello World` example using their search queries, but then I'm not too fixed on this so I'll revert my changes.

**Comment by BolajiOlajide** [comment]
@courier-new @eseliger @limitedmage This is ready for re-review. 
Thanks!

File: client/web/src/enterprise/batches/create/useSearchTemplate.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import { insertQueryIntoLibraryItem, insertNameIntoLibraryItem } from '../batch-spec/yaml-util'
+
+interface UseSearchQueryResult {
+    renderTemplate?: (title: string) => string
+}
+
+const createRenderTemplate = (query: string): ((title: string) => string) => {
+    let template: string
+    template = insertQueryIntoLibraryItem(helloWorldSample, query)
+    template = template.replace('# Find all repositories that contain a README.md file.\n', '')
```
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply client/web/src/enterprise/batches/create/useSearchTemplate.ts:13]: That makes sense to me.

[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply client/web/src/enterprise/batches/create/useSearchTemplate.ts]: CleanShot 2022-06-09 at 22 43 11@2x

Done!

File: client/web/src/enterprise/batches/batch-spec/useBatchSpecCode.ts [hunk 6]
```diff
@@ -133,7 +132,7 @@ export const useBatchSpecCode = (originalInput: string, name: string): UseBatchS
     const [code, setCode] = useState<string>(() =>
         // Start with the hello world sample code initially if the user hasn't written any
         // batch spec code yet, otherwise show the latest spec code.
-        isMinimalBatchSpec(originalInput) ? insertNameIntoLibraryItem(helloWorldSample, name) : originalInput
+        isMinimalBatchSpec(originalInput) ? insertNameIntoLibraryItem(originalInput, name) : originalInput
```
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply client/web/src/enterprise/batches/batch-spec/useBatchSpecCode.ts]: Fixed.

File: client/web/src/enterprise/batches/create/useSearchTemplate.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply client/web/src/enterprise/batches/create/useSearchTemplate.ts]: Fixed.

**Pull request merged**